lib/WUI/http_lifetime.cpp

```diff
@@ -0,0 +1,196 @@
+#include "nhttp/server.h"
+#include "nhttp/common_selectors.h"
+#include "nhttp/headers.h"
+#include "nhttp/static_mem.h"
+#include "nhttp/gcode_upload.h"
+#include "wui.h"
+#include "wui_api.h"
+#include "wui_REST_api.h"
+
+#include <FreeRTOS.h>
+#include <semphr.h>
+#include <lwip/apps/fs.h>
+
+#include <cassert>
+#include <cstring>
+
+using std::nullopt;
+using std::optional;
+using std::string_view;
+using namespace nhttp;
+using namespace nhttp::handler;
+using namespace nhttp::handler::selectors;
+using nhttp::printer::GcodeUpload;
+
+namespace {
+
+SemaphoreHandle_t httpd_mutex = NULL;
+
+// TODO: Move to http_handler_default.cpp?
+// TODO: Eventually remove in favor of files in XFlash.
+class StaticFsFile final : public Selector {
+public:
+    virtual optional<ConnectionState> accept(const RequestParser &parser) const override {
+        char filename[MAX_URL_LEN + 1];
+
+        if (!parser.uri_filename(filename, sizeof(filename))) {
+            return nullopt;
+        }
+
+        if (strcmp(filename, "/") == 0) {
+            strncpy(filename, "/index.html", MAX_URL_LEN);
+        }
+
+        fs_file file;
+        if (fs_open(&file, filename) == ERR_OK) {
+#if LWIP_HTTPD_CUSTOM_FILES
+            // These are not static chunk of memory, not supported.
+            if (file.is_custom_file) {
+                fs_close(&file);
+                return nullopt;
+            }
+#endif
+            /*
+             * The data of non-custom fs files are embedded inside the program
+             * as C arrays. Therefore, we can get the data and close the file
+             * right away.
+             */
+            /*
+             * We guess the content type by a file extension. Looking inside
+             * and doing "file magic" would be better/more reliable, but don't
+             * want the code complexity of that.
+             */
+            static const char *extra_hdrs[] = { "Content-Encoding: gzip\r\n", nullptr };
+            SendStaticMemory send(string_view(file.data, file.len), guess_content_by_ext(filename), parser.can_keep_alive(), extra_hdrs);
```

Comment:
... also in case of sending something else than a static string from CPUFLASH, the `string_view(file.data, file_length)` would have to be changed.

Reply:
In that case we will have a different „Handler“ than `SendStaticMemory`. I have a branch that adds `SendFile`, that gets a `File *` and sends that.

Reply:
ok